### PR TITLE
Support health monitoring mode for NPD

### PR DESCRIPTION
**Pull request by jessieqliu**
Supports the full list of system stat metrics that can be collected by NPD for health monitoring. The full configuration will be enabled if allowed by the policy and spec.

## Breaking Changes
- Changes the signature of `spec.GetLaunchPolicy(imageLabels map[string]string)` to `spec.GetLaunchPolicy(imageLabels map[string]string, logger *log.Logger)`

### CODE REVIEW

File: launcher/spec/launch_spec.go [hunk 1]
```diff
@@ -154,9 +155,22 @@ func (s *LaunchSpec) UnmarshalJSON(b []byte) error {
...
 
-	if val, ok := unmarshaledMap[memoryMonitoringEnable]; ok && val != "" {
-		if boolValue, err := strconv.ParseBool(val); err == nil {
-			s.MemoryMonitoringEnabled = boolValue
+	memVal, memOk := unmarshaledMap[memoryMonitoringEnable]
+	monVal, monOk := unmarshaledMap[monitoringEnable]
+
+	if memOk && monOk {
+		return fmt.Errorf("both %v and %v are specified, only one is permitted", memoryMonitoringEnable, monitoringEnable)
+	} else if memOk && memVal != "" {
```
Comment (alexmwu) [inline launcher/spec/launch_spec.go:163]: I know we didn't do this before, but why don't we fail if ParseBool errors? So it's just false?

Also, regardless the memVal != "" check is not needed.

**Review by alexmwu — APPROVED** [review]
Great work! Mostly minor comments except about the struct to json encoding.

File: launcher/internal/healthmonitoring/nodeproblemdetector/systemstats_config.go [hunk 2]
```diff
@@ -4,40 +4,101 @@ package nodeproblemdetector
...
 // For now we only consider collecting memory related metrics.
 // View the comprehensive configuration details on https://github.com/kubernetes/node-problem-detector/tree/master/pkg/systemstatsmonitor#detailed-configuration-options
 type SystemStatsConfig struct {
-	MemoryStatsConfig memoryStatsConfig `json:"memory"`
-	InvokeInterval    string            `json:"invokeInterval"`
+	CPU            *statsConfig `json:"cpu,omitempty"`
+	Disk           *diskConfig  `json:"disk,omitempty"`
+	Host           *statsConfig `json:"host,omitempty"`
+	Memory         *statsConfig `json:"memory,omitempty"`
+	InvokeInterval string       `json:"invokeInterval,omitempty"`
```
[the remark this answers is not in the document]
Comment (jessieqliu) [reply launcher/internal/healthmonitoring/nodeproblemdetector/systemstats_config.go:42]: As discussed offline - the struct was defined with future proofing in mind (if we enable more granular configs in the future).

File: launcher/spec/launch_policy.go [hunk 3]
```diff
@@ -120,6 +205,22 @@ func GetLaunchPolicy(imageLabels map[string]string) (LaunchPolicy, error) {
 	return launchPolicy, nil
 }
 
+func verifyMonitoringConfig(policy MonitoringType, spec MonitoringType) error {
+	if policy == None {
```
[the remark this answers is not in the document]
Comment (jessieqliu) [reply launcher/spec/launch_policy.go:209]: Noted! In the meantime I adjusted the logic a little to be more readable.

**Pull request merged**